### PR TITLE
Local plugin tracker

### PR DESCRIPTION
**Pull request by kchprog**

### CODE REVIEW

**Review by 1whatleytay — APPROVED** [review]

File: src/main/java/org/hydev/mcpm/client/database/ListAllBoundary.java [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+    public List<String> listBoundary(String parameter) {
+        try {
+            return listAllInteractor.listAll(parameter);
+        } catch (Exception e) {
+            e.printStackTrace();
+            return null;
+        }
+    }
+}
```
Comment (hykilpikonna) [inline src/main/java/org/hydev/mcpm/client/database/ListAllBoundary.java]: Why is this a class? ListAllBoundary should be an interface

Comment (1whatleytay) [reply src/main/java/org/hydev/mcpm/client/database/ListAllBoundary.java]: ~~Truth. This looks like a Controller to me.~~

File: src/main/java/org/hydev/mcpm/client/database/ListAllBoundary.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+import java.util.List;
+import java.util.ArrayList;
+import java.io.*;  
+import java.util.Scanner;
+
+import javax.naming.NameNotFoundException;
+import javax.swing.plaf.metal.MetalIconFactory.FileIcon16;  
+
+public class ListAllBoundary {
+    ListAllInteractor listAllInteractor;
```
Comment (1whatleytay) [inline src/main/java/org/hydev/mcpm/client/database/ListAllBoundary.java]: Take this in the constructor.
```suggestion
    private final ListAllBoundary listAllBoundary;
```

File: src/main/java/org/hydev/mcpm/client/database/ListAllController.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+    /**
+     * Executes the ListAll use case.
+     *
+     * @param parameter The parameter for the ListAll use case.
+     * @return The list of plugins.
+     */
+    List<String> listAll(String parameter) {
+
+        String[] validIn = {"all", "manual", "outdated"};
```
Comment (1whatleytay) [inline src/main/java/org/hydev/mcpm/client/database/ListAllController.java:49]: Set?

File: src/main/java/org/hydev/mcpm/client/database/ListAllInteractor.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+import org.apache.commons.lang3.NotImplementedException;
+import org.hydev.mcpm.*;
+import org.hydev.mcpm.client.models.PluginYml;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class ListAllInteractor {
+    LocalPluginTracker localPluginTracker = new LocalPluginTracker();
```
Comment (1whatleytay) [inline src/main/java/org/hydev/mcpm/client/database/ListAllInteractor.java:17]:
```suggestion
    private final PluginTracker localPluginTracker;
```

Take it as part of the constructor!!

File: src/main/java/org/hydev/mcpm/client/database/LocalPluginTracker.java [hunk 5]
```diff
@@ -55,6 +83,120 @@ private PluginYml readMeta(File jar) {
...
+        }
+    }
+
+
+    /**
+     * Add a plugin to the CSV file
+     * @param name Plugin name
+     * @param status Plugin status (true = manual, false = auto)
+     */
+    public void addEntry(String name, boolean status) {
```
Comment (1whatleytay) [inline src/main/java/org/hydev/mcpm/client/database/LocalPluginTracker.java:113]: Cool!

File: src/main/java/org/hydev/mcpm/client/database/LocalPluginTracker.java [hunk 6]
```diff
@@ -119,48 +232,35 @@ private void updateCsv(String replace, int row, int col) throws IOException {
      * PluginName2,false
      *
      * @param name Plugin name
-     */
-    
-    public void addManuallyInstalled(String name)
+     */    
+    public void setManuallyInstalled(String name)
```
Comment (1whatleytay) [inline src/main/java/org/hydev/mcpm/client/database/LocalPluginTracker.java]: You can merge this method and `removeManuallyInstalled` together by just taking in a `boolean` for whether or not its manual.

File: src/main/java/org/hydev/mcpm/client/database/export/Export.java [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+import org.hydev.mcpm.client.database.fetcher.ProgressBarFetcherListener;
+import org.hydev.mcpm.client.database.inputs.ExportPluginsInput;
+import org.hydev.mcpm.client.database.results.ExportPluginsResult;
+
+import java.io.PrintStream;
+
+/**
+ * An implementation of ExportPluginsBoundary that fetches from a database.
+ */
+public class Export implements ExportPluginsBoundary {
```
Comment (1whatleytay) [inline src/main/java/org/hydev/mcpm/client/database/export/Export.java]: Wait wut is this doing here...

File: src/main/java/org/hydev/mcpm/client/database/ListAllBoundary.java [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+
+    public List<String> listBoundary(String parameter) {
+        try {
+            return listAllInteractor.listAll(parameter);
+        } catch (Exception e) {
+            e.printStackTrace();
+            return null;
```
Comment (hykilpikonna) [inline src/main/java/org/hydev/mcpm/client/database/ListAllBoundary.java]: Maybe don't catch all `Exception`

**Review by 1whatleytay — APPROVED** [review]
Still some weird stuff (export + you still use LocalPluginTracker). If this can't be fixed now then I guess we should merge.

**Review by hykilpikonna — APPROVED** [review]

**Pull request merged**